### ISSUE TITLE
rpc: add an error type for custom errors

### ISSUE THREAD

**Issue by psarna** [scylladb/seastar#835]
Currently, there are two types of propagated errors:
 * user error (`USER`)
 * unknown RPC verb (`UNKNOWN_VERB`)

The user error is used to propagate all kinds of errors via rpc, but all of them translate to `std::runtime_error` with a custom message. That's theoretically enough to recognize specific error types, but parsing error strings in order to make decisions about the flow is bad practice.

I propose adding another error type `USER_TYPED`, which has a 32-bit `type` field and a byte blob with custom payload (either just a message or some more complicated context specific to the error type). That would allow to easily distinguish what's the error type and act accordingly. In particular, an rpc client could be optionally created with a custom map which stores this 64-bit keys and concrete c++ exceptions as values, and throws an appropriate error when received (somewhere in `unmarshal_exception` I suppose). I specifically want to use such a mechanism later to allow propagating `OverloadedException` from Scylla replicas to the coordinator.

@gleb-cloudius I'd like to hear your opinion on this before starting to code, does it make sense? If it does, I'll prepare a seastar series first and then use it to solve the Scylla issue.

**Comment by gleb-cloudius** [comment]
On Fri, Dec 04, 2020 at 08:52:26AM -0800, Piotr Sarna wrote:
> Currently, there are two types of propagated errors:
>  * user error (`USER`)
>  * unknown RPC verb (`UNKNOWN_VERB`)
> 
> The user error is used to propagate all kinds of errors via rpc, but all of them translate to `std::runtime_error` with a custom message. That's theoretically enough to recognize specific error types, but parsing error strings in order to make decisions about the flow is bad practice.
> 
> I propose adding another error type `USER_TYPED`, which has a 32-bit `type` field and a byte blob with custom payload (either just a message or some more complicated context specific to the error type). That would allow to easily distinguish what's the error type and act accordingly.
> In particular, an rpc client could be optionally created with a custom map which stores this 64-bit
I guess you meant 32-bit here.

> keys and concrete c++ exceptions as values, and throws an appropriate error when received (somewhere in `unmarshal_exception` I suppose). I specifically want to use such a mechanism later to allow propagating `OverloadedException` from Scylla replicas to the coordinator.
> 
> @gleb-cloudius I'd like to hear your opinion on this before starting to code, does it make sense? If it does, I'll prepare a seastar series first and then use it to solve the Scylla issue.
> 
We discussed passing typed exceptions via rpc before. The reason it was
not done is because it is not clear how to tell in IDL what exceptions
a call may return. In rpc the function signature is the idl, so suppose
we have an rpc call:   float divide(float x, float y). How can we figure
out from this signature that it can return divide by zero exception?
Unlike in Java in C++ there is not info about exception in a function
signature.  You propose to create a global exception map for each client,
but errors are property of a call, not a client. Also it is not clear
how compatibility will be handled. What if exception is added but a
caller is still old and gets an exception not from the map?

OTOH all of those are easily addressable by communicating errors as
values over RPC. So if you want to return OverloadedException make your
rpc call be:

std::variant<OverloadedException, mutation> read(...).


--
			Gleb.

**Comment by psarna** [comment]
Alright, `std::variant` is one way, but I'd rather avoid it for a couple reasons:
 * creating a second way of propagating errors, even though there is one already and it supports "user" errors
 * wasting another bit sent via network, while we already have bits available for use
 * introducing inevitable backward compatibility mess to Scylla

Would you be opposed to introducing a generic `OVERLOADED` error type to Seastar RPC then? It could have very clear semantics and a separate error thrown when it happens (`rpc::overloaded`), and it's quite safe to assume that it will be useful outside of Scylla - Seastar projects are assumed to be high-performance servers, which means they could almost surely use a mechanism for propagating overload errors via RPC for their internal overload protection schemes. I wasn't initially a fan of this idea, but the more I think about it the more sense it makes, since the RPC layer *can* be overloaded with too many messages, so it may also be able to communicate a specific error type in such situations.

**Comment by gleb-cloudius** [comment]
On Sun, Dec 06, 2020 at 10:44:26PM -0800, Piotr Sarna wrote:
> Alright, `std::variant` is one way, but I'd rather avoid it for a couple reasons:
>  * creating a second way of propagating errors, even though there is one already and it supports "user" errors
User error is a catch all case because we have to have one or require rpc
handlers to be noexcept. On a hindsight we should have done the later.
I do not think there are RPCs out there that support propagating
exceptions from handlers. Doing so will only complicate their IDL
without providing any benefits over simple return value. gRPC for
instance returns GRPC_STATUS_UNKNOWN if a handler throws.

>  * wasting another bit sent via network, while we already have bits available for use
Byte, not bit. This will not kill us.

>  * introducing inevitable backward compatibility mess to Scylla
> 
Changing rpc protocol will do that. Introducing another verb with
different return value OTOH is well established procedure by now and
will only affect verbs that needs it.

BTW how are you going handle one way verbs like MUTATE that do not
return any value at all. 


> Would you be opposed to introducing a generic `OVERLOADED` error type to Seastar RPC then? It could have very clear semantics and a separate error thrown when it happens (`rpc::overloaded`), and it's quite safe to assume that it will be useful outside of Scylla - Seastar projects are assumed to be high-performance servers, which means they could almost surely use a mechanism for propagating overload errors via RPC for their internal overload protection schemes. I wasn't initially a fan of this idea, but the more I think about it the more sense it makes, since the RPC layer *can* be overloaded with too many messages, so it may also be able to communicate a specific error type in such situations.
> 
There is a difference between rpc overload and an overload of a server
that uses rpc. I have no problem to introduce rpc::overload for the
former if there is a use case for it (the capability has to be
negotiated of course), but it should not be used for the later.

--
			Gleb.

**Comment by psarna** [comment]
On 12/7/20 8:23 AM, Gleb Natapov wrote:
> On Sun, Dec 06, 2020 at 10:44:26PM -0800, Piotr Sarna wrote:
> > Alright, `std::variant` is one way, but I'd rather avoid it for a 
> couple reasons:
> > * creating a second way of propagating errors, even though there is 
> one already and it supports "user" errors
> User error is a catch all case because we have to have one or require rpc
> handlers to be noexcept. On a hindsight we should have done the later.
> I do not think there are RPCs out there that support propagating
> exceptions from handlers. Doing so will only complicate their IDL
> without providing any benefits over simple return value. gRPC for
> instance returns GRPC_STATUS_UNKNOWN if a handler throws.
>
> > * wasting another bit sent via network, while we already have bits 
> available for use
> Byte, not bit. This will not kill us.
>
> > * introducing inevitable backward compatibility mess to Scylla
> >
> Changing rpc protocol will do that. 
Doesn't it simply boil down to agreeing whether to use or not use this 
type of error during rpc negotiation?
> Introducing another verb with
> different return value OTOH is well established procedure by now and
> will only affect verbs that needs it.
It is, but then we have two ways of transporting user errors - one in 
the rpc layer, used for most things, and one for overloaded errors, used 
one layer up. And later, if we want to propagate yet another type of 
error, the discrepancy will simply get bigger.
>
> BTW how are you going handle one way verbs like MUTATE that do not
> return any value at all.

For now I leave it as Scylla's problem, it was a design decision to not 
expect any value at all and it can be changed if needed. In short term I 
only need to propagate overloads from reader concurrency semaphores, 
which makes MUTATE temporarily out of scope.

>
>
> > Would you be opposed to introducing a generic `OVERLOADED` error 
> type to Seastar RPC then? It could have very clear semantics and a 
> separate error thrown when it happens (`rpc::overloaded`), and it's 
> quite safe to assume that it will be useful outside of Scylla - 
> Seastar projects are assumed to be high-performance servers, which 
> means they could almost surely use a mechanism for propagating 
> overload errors via RPC for their internal overload protection 
> schemes. I wasn't initially a fan of this idea, but the more I think 
> about it the more sense it makes, since the RPC layer *can* be 
> overloaded with too many messages, so it may also be able to 
> communicate a specific error type in such situations.
> >
> There is a difference between rpc overload and an overload of a server
> that uses rpc. I have no problem to introduce rpc::overload for the
> former if there is a use case for it (the capability has to be
> negotiated of course), but it should not be used for the later.
I understand, but I just see an inconsistency with the `USER` error 
then. It's also something that happened on the server that uses rpc, and 
not property of rpc itself. I agree that `OVERLOAD` is too generic a 
name and could suggest that it's the rpc layer that got overloaded, so 
how about `USER_OVERLOADED`? I just want to have a way of making the 
`USER` error a little bit more specific for a quite generic use case. As 
it is now, to be fully compliant with the idea that only rpc-specific 
errors are transported via rpc exception encoding is to drop `USER` and 
make everyone use `std::variant<X, std::runtime_error>` if they expect 
the user code to fail.

**Comment by gleb-cloudius** [comment]
On Sun, Dec 06, 2020 at 11:37:31PM -0800, Piotr Sarna wrote:
> On 12/7/20 8:23 AM, Gleb Natapov wrote:
> > On Sun, Dec 06, 2020 at 10:44:26PM -0800, Piotr Sarna wrote:
> > > Alright, `std::variant` is one way, but I'd rather avoid it for a 
> > couple reasons:
> > > * creating a second way of propagating errors, even though there is 
> > one already and it supports "user" errors
> > User error is a catch all case because we have to have one or require rpc
> > handlers to be noexcept. On a hindsight we should have done the later.
> > I do not think there are RPCs out there that support propagating
> > exceptions from handlers. Doing so will only complicate their IDL
> > without providing any benefits over simple return value. gRPC for
> > instance returns GRPC_STATUS_UNKNOWN if a handler throws.
> >
> > > * wasting another bit sent via network, while we already have bits 
> > available for use
> > Byte, not bit. This will not kill us.
> >
> > > * introducing inevitable backward compatibility mess to Scylla
> > >
> > Changing rpc protocol will do that. 
> Doesn't it simply boil down to agreeing whether to use or not use this 
> type of error during rpc negotiation?
I was talking about error map that you proposed before. How do you know
what errors are supported or not supported there. Just adding a certain
type of error into the protocol itself can be negotiated by rpc.

> > Introducing another verb with
> > different return value OTOH is well established procedure by now and
> > will only affect verbs that needs it.
> It is, but then we have two ways of transporting user errors - one in 
> the rpc layer, used for most things, and one for overloaded errors, used 
> one layer up. And later, if we want to propagate yet another type of 
> error, the discrepancy will simply get bigger.
The rpc one should not be considered as a generic way to transport errors,
just a fact that a handler execution failed. If you want to transport
meaningful error make it a return value.

> >
> > BTW how are you going handle one way verbs like MUTATE that do not
> > return any value at all.
> 
> For now I leave it as Scylla's problem, it was a design decision to not 
> expect any value at all and it can be changed if needed.
It is not that easy to change it since there is no direct communication
with all hosts. The verb is forwarded by only one node into other DCs.

>                                                           In short term I 
> only need to propagate overloads from reader concurrency semaphores, 
> which makes MUTATE temporarily out of scope.
> 
> >
> >
> > > Would you be opposed to introducing a generic `OVERLOADED` error 
> > type to Seastar RPC then? It could have very clear semantics and a 
> > separate error thrown when it happens (`rpc::overloaded`), and it's 
> > quite safe to assume that it will be useful outside of Scylla - 
> > Seastar projects are assumed to be high-performance servers, which 
> > means they could almost surely use a mechanism for propagating 
> > overload errors via RPC for their internal overload protection 
> > schemes. I wasn't initially a fan of this idea, but the more I think 
> > about it the more sense it makes, since the RPC layer *can* be 
> > overloaded with too many messages, so it may also be able to 
> > communicate a specific error type in such situations.
> > >
> > There is a difference between rpc overload and an overload of a server
> > that uses rpc. I have no problem to introduce rpc::overload for the
> > former if there is a use case for it (the capability has to be
> > negotiated of course), but it should not be used for the later.
> I understand, but I just see an inconsistency with the `USER` error 
> then. It's also something that happened on the server that uses rpc, and 
> not property of rpc itself.
It is! It communicates the fact that execution of use handler failed (as
opposite of not been found for instance).

>                              I agree that `OVERLOAD` is too generic a 
> name and could suggest that it's the rpc layer that got overloaded, so 
> how about `USER_OVERLOADED`? I just want to have a way of making the 
> `USER` error a little bit more specific for a quite generic use case. As 
> it is now, to be fully compliant with the idea that only rpc-specific 
> errors are transported via rpc exception encoding is to drop `USER` and 
> make everyone use `std::variant<X, std::runtime_error>` if they expect 
> the user code to fail.
Your view of USER error is incorrect. It is not a way for a handler to
return an error, it is a way for an RPC server to communicate the fact
that a handler did not return a value.

I do not think high (above rpc) level protocol errors should have
special treatment in rpc protocol itself. RPC already has a generic way
to allow your verb to transport any information it wants. It may be a
variant, it may be a status field in a structure been returned, it may
be a optional<value> where disengaged one means an error.

If you add errors to RPC protocol itself then RPC is the one that
defines what the error means and what the error carries. What if you
want to extend the overload error with additional information later on
(queue size or whatever)? If you define it in your IDL you can do it in
backwards compatible way without change rpc at all, if it is in the rpc
you need to change the rpc and negotiate new features.

For read I think it can be even done in backwards compatible way by
making a verb returning rpc::tuple<value, status> and returning empty
value with appropriate status in case of an overload. Variant is cleaner
of course but can't be achieved without introducing new verb.

--
			Gleb.

**Comment by psarna** [comment]
Ok. This issue is no longer relevant so I'm closing it.
